rust-src/curve_arithmetic/src/bls12_381_g1hash.rs

```diff
@@ -496,41 +485,139 @@ pub(crate) const K4: [[u64; 6]; 16] = [
-    p0.add_assign(&p1);
-
-    // Clear cofactors (ensuring that the point is in the correct order subgroup of
-    // the curve)
-    p0.mul_assign(1 + 15_132_376_222_941_642_752);
-    p0
+/// Implements https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-10#section-3
+/// It follows the steps
+///    1. u = hash_to_field(msg, 2)
+///    2. Q0 = map_to_curve(u[0])
+///    3. Q1 = map_to_curve(u[1])
+///    4. R = Q0 + Q1              
+///    5. P = clear_cofactor(R) = h_eff * R   # Clearing cofactor
+///    6. return P,
+/// where the choices of hash_to_field, map_to_curve and h_eff are as described in https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-10#section-8.8.1.
+pub fn hash_to_curve(msg: &[u8]) -> G1 {
+    let dst = "QUUX-V01-CS02-with-BLS12381G1_XMD:SHA-256_SSWU_RO_".as_bytes();
+    let (u0, u1) = hash_to_field(msg, dst);
+
+    let q0 = map_to_curve(u0); // This is on E, but not necessarily in G1
+    let q1 = map_to_curve(u1); // This is on E, but not necessarily in G1
+
+    let mut r = q0;
+    r.add_assign(&q1); // This is on E, but not necessarily in G1
+    r.mul_assign(15132376222941642753); // Clearing cofactor with h_eff = 15132376222941642753
+    r // This now guarantied to be in G1
+}
+
+/// Implements https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-10#section-6.6.3
+/// It uses the function sswu_3mod4 to get xn, xd, y such that (xn/xd : y : 1)
+/// is a point on E' (written in Jacobian coordinates). Since inversions are
+/// expensive, we use the fact that (xn/xd : y : 1) on E' <=> (xn xd : xd^3y :
+/// xd) on E' when using Jacobian coordinates. It returns iso_11((xn xd : xd^3y
+/// : xd)) which is then guarentied to lie on E (but not necessarily in G1).
+fn map_to_curve(u: Fq) -> G1 {
+    let (xn, xd, mut y, _) = sswu_3mod4(u);
+    let mut x = xn;
+    x.mul_assign(&xd);
+    y.mul_assign(&xd);
+    y.mul_assign(&xd);
+    y.mul_assign(&xd);
+    let (xiso, yiso, z) = iso_11(x, y, xd);
+    from_coordinates_unchecked(xiso, yiso, z).expect("should not happen")
+}
+
+/// Implements https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-10#section-3
+/// with the choice of expand_message being expand_message_xmd, as specified in https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-10#section-8.8.1.
+fn hash_to_field(msg: &[u8], dst: &[u8]) -> (Fq, Fq) {
+    let (u_0, u_1, u_2, u_3) = expand_message_xmd(msg, dst);
+
+    (fq_from_bytes(&u_0, &u_1), fq_from_bytes(&u_2, &u_3))
+}
+
+// Interpret input as integers (big endian)
+// Return (left*2^256 + right) as Fq
+fn fq_from_bytes(left_bytes: &[u8; 32], right_bytes: &[u8; 32]) -> Fq {
+    fn le_u64s_from_be_bytes(bytes: &[u8]) -> Fq {
+        let mut digits = [0u64; 6];
+
+        for i in 0..4 {
+            digits[3 - i] = u64::from_be_bytes(
+                bytes[i * 8..(i + 1) * 8]
+                    .try_into()
+                    .expect("Chunk Size is always 8."),
+            );
+        }
```

Comment:
I think it'd be clearer to use the `chunks` method here like we do when deserializing field elements.

```rust
for (place, chunk) in digits.iter_mut().zip(bytes.chunks(8).rev()) {
    *place = u64::from_be_bytes(chunk.try_into().expect("Chunk Sie is always 8"))
}
```

Reply:
But you decide.